security/s2a/internal/crypter/aesgcm_test.go

```diff
@@ -0,0 +1,474 @@
+	}
+}
+
+// Test update key for AES-GCM using a key with different size from the initial
+// key.
+func TestAESGCMKeySizeUpdate(t *testing.T) {
+	for _, tc := range []struct {
+		desc          string
+		updateKeySize int
+	}{
+		{"mismatch key size update", aes256GcmKeySize},
+		{"invalid key size update", 17},
+	} {
+		t.Run(tc.desc, func(t *testing.T) {
+			key := make([]byte, aes128GcmKeySize)
+			crypter, err := NewAESGCM(key)
+			if err != nil {
+				t.Fatalf("NewAESGCM(keySize=%v) failed, err: %v", aes128GcmKeySize, err)
+			}
+
+			// Update the key with a new one which is a different from the original.
+			newKey := make([]byte, tc.updateKeySize)
+			if err = crypter.UpdateKey(newKey); err == nil {
+				t.Fatal("UpdateKey should fail with invalid key size error")
+			}
+		})
+	}
+}
+
+// Test Encrypt/Decrypt using an invalid nonce size.
+func TestAESGCMEncryptDecryptInvalidNonce(t *testing.T) {
+	key := make([]byte, aes128GcmKeySize)
+	crypter, err := NewAESGCM(key)
+	if err != nil {
+		t.Fatalf("NewAESGCM(keySize=%v) failed, err: %v", aes128GcmKeySize, err)
+	}
+	// Construct nonce with invalid size.
+	nonce := make([]byte, 1)
+	if _, err = crypter.Encrypt(nil, nil, nonce, nil); err == nil {
+		t.Errorf("Encrypt should fail due to invalid nonce size")
+	}
+	if _, err = crypter.Decrypt(nil, nil, nonce, nil); err == nil {
+		t.Fatalf("Decrypt should fail due to invalid nonce size")
+	}
+}
+
+// Test encrypt and decrypt on roundtrip messages for AES-GCM with and without
+// updating the keys.
+func TestAESGCMEncryptRoundtrip(t *testing.T) {
+	for _, shouldUpdateKeys := range []bool{true, false} {
+		for _, keySize := range []int{aes128GcmKeySize, aes256GcmKeySize} {
+			key := make([]byte, keySize)
+			sender, receiver := getGCMCryptoPair(key, t)
+			if shouldUpdateKeys {
+				// Update the key with a new one which is different from the original.
+				newKey := make([]byte, keySize)
+				newKey[0] = '\xbd'
+				if err := sender.UpdateKey(newKey); err != nil {
+					t.Fatalf("sender UpdateKey failed with: %v", err)
+				}
+				if err := receiver.UpdateKey(newKey); err != nil {
+					t.Fatalf("receiver UpdateKey failed with: %v", err)
+				}
+			}
+			testGCMEncryptRoundtrip(sender, receiver, t)
+		}
+	}
+}
+
+// Test encrypt and decrypt using test vectors for aes128gcm. Much of these
+// test vectors were taken from the ALTS AES-GCM unit tests.
+func TestAESGCMEncrypt(t *testing.T) {
+	for _, test := range []testutil.CryptoTestVector{
+		{
+			Key:         testutil.Dehex("5b9604fe14eadba931b0ccf34843dab9"),
+			Plaintext:   testutil.Dehex("001d0c231287c1182784554ca3a21908"),
+			Ciphertext:  testutil.Dehex("26073cc1d851beff176384dc9896d5ff"),
+			Tag:         testutil.Dehex("0a3ea7a5487cb5f7d70fb6c58d038554"),
+			Nonce:       testutil.Dehex("028318abc1824029138141a2"),
+			Result:      testutil.ValidResult,
+			AllocateDst: true,
+		},
+		{
+			Key:         testutil.Dehex("11754cd72aec309bf52f7687212e8957"),
+			Plaintext:   nil,
```

Comment:
You can delete any field that has the default value. Some default values are: nil, 0, "", false.

Reply:
Done, though I think some explicit values may help with determining what exact case the test vector is testing.

Reply:
This is why we added `Comment` :)